### PR TITLE
WIP Test Framework

### PR DESCRIPTION
**Pull request by JPercival**
In starting to build the submit-data tests and reviewing the work that's been done so far it became obvious that I hadn't provided quite enough guidance on testing or on building out shared components. This is a PR to fix a few of those things:

* Reworked/streamlined utility classes, and guidance on how to create them

Read the documentation in the base README

A concrete example for usage of some of the new classes:

```java
// Old way
private IBundleProvider performCodeSystemSearchByUrl(String rxNormUrl) {
  return this.myDaoRegistry.getResourceDao(CodeSystem.class)
		.search(SearchParameterMap.newSynchronous().add(CodeSystem.SP_URL, new UriParam(rxNormUrl)));
}

// New way
private IBundleProvider performCodeSystemSearchByUrl(String rxNormUrl) {
   return search(CodeSystem.class, Searches.byUrl(rxNormUrl));
}
```

* OperationProvider base class that is set up to provide easy local CRUD
* Base classes for Rest (meaning you need to spin the whole server up) and Dao (meaning you only need the database) Integration tests (WIP at the time of this writing, new to share spring properties so they aren't copy and pasted everywhere)
  * Dao only tests start about twice as fast, so nice to use them to validate operation provider behavior. They are not sufficient to validate that an operation provider is registered on a server
* Reworked/streamlined test case loading
* Code demonstrating how to check preconditions on public apis
* Usage of the most specific assertion type for tests

```java
// Yes
assertEquals(thisThing, thatThing)

// No
assertTrue(this == that)
```
* Fixes for incorrect assertion ordering

Still to do:
* Tests and docs for new code
* Fixing broken tests

### CODE REVIEW

File: common/src/main/java/org/opencds/cqf/ruler/common/provider/DaoRegistryOperationProvider.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import org.opencds.cqf.ruler.api.OperationProvider;
+import org.opencds.cqf.ruler.common.utility.DaoRegistryUser;
+import org.opencds.cqf.ruler.common.utility.FhirContextUser;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.jpa.api.dao.DaoRegistry;
+
+public class DaoRegistryOperationProvider implements OperationProvider, DaoRegistryUser, FhirContextUser {
```
Comment (JPercival) [inline common/src/main/java/org/opencds/cqf/ruler/common/provider/DaoRegistryOperationProvider.java:11]: This is a new class that provides CRUD helpers out of the box.

File: common/src/main/java/org/opencds/cqf/ruler/common/utility/Canonicals.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		String getVersion();
+
+		String getUrl();
+
+		IdType getId();
+	}
+
+	public static <CanonicalType extends IPrimitiveType<String>> String getId(CanonicalType theCanonicalType) {
+		checkNotNull(theCanonicalType);
+		checkArgument(theCanonicalType.getValue() != null);
```
Comment (JPercival) [inline common/src/main/java/org/opencds/cqf/ruler/common/utility/Canonicals.java:26]: These are examples of validating inputs to public APIs. Validate inputs as the very first step in a function.

File: common/src/main/java/org/opencds/cqf/ruler/common/utility/DaoRegistryUser.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import ca.uhn.fhir.jpa.api.dao.IFhirResourceDao;
+import ca.uhn.fhir.jpa.api.model.DaoMethodOutcome;
+import ca.uhn.fhir.jpa.searchparam.SearchParameterMap;
+import ca.uhn.fhir.rest.api.server.IBundleProvider;
+import ca.uhn.fhir.rest.api.server.RequestDetails;
+
+/**
+ * Simulate FhirDal operations until such time as that is fully baked
+ */
+public interface DaoRegistryUser {
```
Comment (JPercival) [inline common/src/main/java/org/opencds/cqf/ruler/common/utility/DaoRegistryUser.java:19]: We're working through an implementation of a common API for Java FHIR access. This interface approximates that so it's easier to get convergence when that finally happens.

File: common/src/main/java/org/opencds/cqf/ruler/common/utility/ResourceCreator.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.instance.model.api.IIdType;
+
+public interface ResourceCreator extends FhirContextUser {
+	@SuppressWarnings("unchecked")
+	default <T extends IBaseResource, I extends IIdType> T newResource(I theId) {
```
Comment (JPercival) [inline common/src/main/java/org/opencds/cqf/ruler/common/utility/ResourceCreator.java:11]: This is a helper that lets you create a new resource of the correct type and version based only on the id.

File: common/src/main/java/org/opencds/cqf/ruler/common/utility/Searches.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import ca.uhn.fhir.jpa.searchparam.SearchParameterMap;
+import ca.uhn.fhir.model.api.IQueryParameterType;
+import ca.uhn.fhir.rest.param.StringParam;
+import ca.uhn.fhir.rest.param.UriParam;
+
+public class Searches {
+	private Searches() {
+	}
+
+	public static SearchParameterMap all() {
```
Comment (JPercival) [inline common/src/main/java/org/opencds/cqf/ruler/common/utility/Searches.java:14]: Find all resources (when combined with the Dao API above)

```java
search(Measure.class, Searches.all())
```

File: common/src/main/java/org/opencds/cqf/ruler/common/utility/Searches.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package org.opencds.cqf.ruler.common.utility;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import ca.uhn.fhir.jpa.searchparam.SearchParameterMap;
+import ca.uhn.fhir.model.api.IQueryParameterType;
+import ca.uhn.fhir.rest.param.StringParam;
+import ca.uhn.fhir.rest.param.UriParam;
+
+public class Searches {
```
Comment (JPercival) [inline common/src/main/java/org/opencds/cqf/ruler/common/utility/Searches.java:10]: This is a set of utilities to construct commonly needed search parameter maps.

File: plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java [hunk 7]
```diff
@@ -61,17 +56,12 @@ public Bundle extractLineListData(RequestDetails details,
...
 		Map<String, Bundle> eicrs = new HashMap<String, Bundle>();
 		for (Map.Entry<String, Reference> entry : populationSubjectListReferenceMap.entrySet()) {
-			// IQueryParameterType compositionReferenceParam = new ReferenceParam("subject", entry.getKey());
-			SearchParameterMap map = SearchParameterMap.newSynchronous();//.add("composition", compositionReferenceParam);
-			logger.warn(map.toNormalizedQueryString(measureResourceProvider.getContext()));
-			List<IBaseResource> bundles = myDaoRegistry.getResourceDao(Bundle.class).search(map).getAllResources();
-			for (IBaseResource baseBundle : bundles) {
-				if (baseBundle instanceof Bundle) {
-					Bundle bundle = (Bundle) baseBundle;
+			List<Bundle> bundles = search(Bundle.class, Searches.all()).getAllResourcesTyped();
```
Comment (JPercival) [inline plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java:59]: This line combines the Search API with the new TypedBundleProvider. Compare to the previous code.

File: plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java [hunk 8]
```diff
@@ -125,11 +112,7 @@ private void gatherPatientsFromReport(MeasureReport report, Map<String, Referenc
 
 	private List<Reference> getPatientListFromGroup(String subjectGroupRef) {
 		List<Reference> patientList = new ArrayList<>();
-		IBaseResource baseGroup = myDaoRegistry.getResourceDao("Group").read(new IdType(subjectGroupRef));
-		if (baseGroup == null) {
-			throw new RuntimeException("Could not find Group/" + subjectGroupRef);
-		}
-		Group group = (Group) baseGroup;
+		Group group = read(new IdType(subjectGroupRef));
```
Comment (JPercival) [inline plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java:115]: Here's another example to compare to the previous code.

File: plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java [hunk 9]
```diff
@@ -149,9 +132,9 @@ private void gatherPatientsFromReport(MeasureReport report, Map<String, Referenc
 			logger.debug("No subject results found.");
 			return results;
 		}
-		IBaseResource baseList = myDaoRegistry.getResourceDao(subjectResults.getReferenceElement().getResourceType()).read(subjectResults.getReferenceElement());
+		IBaseResource baseList = read(subjectResults.getReferenceElement());
```
Comment (JPercival) [inline plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java:135]: And one more example to compare to the old code.

File: plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java [hunk 10]
```diff
@@ -149,9 +132,9 @@ private void gatherPatientsFromReport(MeasureReport report, Map<String, Referenc
...
 		if (baseList == null) {
-			logger.debug(String.format("No subject results found for: %s", subjectResults.getReference()));
+			logger.debug("No subject results found for: {}", subjectResults.getReference());
```
Comment (JPercival) [inline plugin/case-reporting/src/main/java/org/opencds/cqf/ruler/casereporting/r4/MeasureDataProcessProvider.java:137]: There was quite a bit of this too. The loggers accept a string format and parameters, so no need to pre-format it.

File: plugin/cr/src/test/java/org/opencds/cqf/ruler/cr/r4/provider/SubmitDataProviderIT.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+@SpringBootTest(classes = { DaoOnlyConfig.class }, properties = {
+				"scheduling_disabled=true",
+            "spring.main.allow-bean-definition-overriding=true",
+            "spring.batch.job.enabled=false",
+            "hapi.fhir.fhir_version=r4",
+				"hapi.fhir.allow_external_references=true",
+				"hapi.fhir.enforce_referential_integrity_on_write=false",
+})
+@EnableAutoConfiguration(exclude=QuartzAutoConfiguration.class)
+public class SubmitDataProviderIT extends DaoIntegrationTest implements IdCreator, ResourceCreator {
```
Comment (JPercival) [inline plugin/cr/src/test/java/org/opencds/cqf/ruler/cr/r4/provider/SubmitDataProviderIT.java:33]: I'm working on collapsing all the cruft from above into the base classes. Then the body of the tests will become the focus of the code.

File: plugin/cr/src/test/java/org/opencds/cqf/ruler/cr/r4/provider/SubmitDataProviderIT.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	
+	@Autowired
+	SubmitDataProvider mySubmitDataProvider;
+
+	@Test
+	public void testSubmitData() {
```
Comment (JPercival) [inline plugin/cr/src/test/java/org/opencds/cqf/ruler/cr/r4/provider/SubmitDataProviderIT.java:39]: An example of using a few of the utilities to round-trip the database in just a few lines of code.

**Review by rob-reynolds — APPROVED** [review]
There's a lot here to fully digest.  Will adopt patterns in future implementations.

**Pull request merged**